coordinator/src/main/scala/filodb.coordinator/queryplanner/LongTimeRangePlanner.scala

```diff
@@ -180,31 +180,58 @@ import filodb.query.exec._
       case lp: SeriesKeysByFilters         => rawClusterMaterialize(qContext, lp)
-      case lp: ApplyMiscellaneousFunction  => materializeApplyMiscellaneousFunction(qContext, lp)
-      case lp: ApplySortFunction           => materializeApplySortFunction(qContext, lp)
-      case lp: ScalarVaryingDoublePlan     => materializeScalarPlan(qContext, lp)
-      case lp: ScalarTimeBasedPlan         => materializeScalarTimeBased(qContext, lp)
-      case lp: VectorPlan                  => materializeVectorPlan(qContext, lp)
-      case lp: ScalarFixedDoublePlan       => materializeFixedScalar(qContext, lp)
-      case lp: ApplyAbsentFunction         => materializeAbsentFunction(qContext, lp)
-      case lp: ScalarBinaryOperation       => materializeScalarBinaryOperation(qContext, lp)
+      case lp: ApplyMiscellaneousFunction  => super.materializeApplyMiscellaneousFunction(qContext, lp)
+      case lp: ApplySortFunction           => super.materializeApplySortFunction(qContext, lp)
+      case lp: ScalarVaryingDoublePlan     => super.materializeScalarPlan(qContext, lp)
+      case lp: ScalarTimeBasedPlan         => super.materializeScalarTimeBased(qContext, lp)
+      case lp: VectorPlan                  => super.materializeVectorPlan(qContext, lp)
+      case lp: ScalarFixedDoublePlan       => super.materializeFixedScalar(qContext, lp)
+      case lp: ApplyAbsentFunction         => super.materializeAbsentFunction(qContext, lp)
+      case lp: ScalarBinaryOperation       => super.materializeScalarBinaryOperation(qContext, lp)
       case lp: SubqueryWithWindowing       => materializePeriodicSeriesPlan(qContext, lp)
-      case lp: TopLevelSubquery            => materializeTopLevelSubquery(qContext, lp)
+      case lp: TopLevelSubquery            => super.materializeTopLevelSubquery(qContext, lp)
       case lp: ApplyLimitFunction          => rawClusterMaterialize(qContext, lp)
       case lp: LabelNames                  => rawClusterMaterialize(qContext, lp)
       case lp: LabelCardinality            => materializeLabelCardinalityPlan(lp, qContext)
     }
     // scalastyle:on cyclomatic.complexity
   }
 
+  override def materializeBinaryJoin(qContext: QueryContext, logicalPlan: BinaryJoin): PlanResult =
+    if (LogicalPlanUtils.getOffsetMillis(logicalPlan).foldLeft(0L)(_.max(_)) == 0) {
+      // No offset, can push down binary join to LT and Raw planner and stitch
+      // Binary joins can be pushed down to respective planners, exporting results to QS is not required
+      val (startTime, endTime) = (logicalPlan.startMs, logicalPlan.endMs)
+      val execPlan = if (startTime > latestDownsampleTimestampFn) {
+        rawClusterPlanner.materialize(logicalPlan, qContext)
+      } else if (endTime < earliestRawTimestampFn) {
+        downsampleClusterPlanner.materialize(logicalPlan, qContext)
+      } else {
+        val dsLogicalPlan = copyLogicalPlanWithUpdatedTimeRange(logicalPlan,
```

Comment:
this is the trickiest part, in general there can be anything under the binary join, for example subquery with overlapping window. Splitting subquery into two expressions one stitching them potentially can bring back invalid results.  This is why I was excited originally when we stopped splitting/stitching high level expressions such as binary join and started doing them on the leave raw series. Stitching raw series from downsample and raw clusters can always be done perfectly and every single expression on top of it would be calculated properly. In that case push down of the aggregation is not possible but the logic is simple and correct. That brings me to the point that probably you meant lookback, not the offset in the above "if" statement. If you read documentation of the LogicalPlanUtils (getLookBackMillis), then this might be what you want (instead  of getOffsetMillis). So, if we don't have aggregation intervals (lookbacks) then it's safe to split the plan. The issue of stitching across the time is the major headache for the subqueries in general. We could do the splits fairly efficiently because there should be an overlap between the downsample and raw cluster and IF this overlap is bigger than the lookback we can still calculate perfectly AND push down aggregation.

Reply:
Can you please provide a concrete PromQL expression for your scenario. That would be helpful for more discussion.

Reply:
avg_over_time(rate(foo[1d])[10d:1h])+ avg_over_time(rate(bar[1d])[10d:1h])
here we have a sliding window of 1d over 10 days every hour. Lookback should be 1 day here. 
It would be interesting to see what logic above would do to this query? What would be the start/end here? I assume it will probably send it all the the raw cluster if "now" is in the raw cluster range.